charger/kathrein.go

```diff
@@ -0,0 +1,635 @@
+func (wb *Kathrein) CurrentPower() (float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(kathreinRegTotalActivePower, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(math.Float32frombits(binary.BigEndian.Uint32(b))), err
+}
+
+var _ api.PhaseCurrents = (*Kathrein)(nil)
+
+// Currents implements the api.PhaseCurrents interface
+func (wb *Kathrein) Currents() (float64, float64, float64, error) {
+	return wb.getPhaseValues(kathreinRegCurrents, 1)
+}
+
+var _ api.PhaseVoltages = (*Kathrein)(nil)
+
+// Voltages implements the api.PhaseVoltages interface
+func (wb *Kathrein) Voltages() (float64, float64, float64, error) {
+	return wb.getPhaseValues(kathreinRegVoltages, 1)
+}
+
+var _ api.PhasePowers = (*Kathrein)(nil)
+
+// Powers implements the api.PhasePowers interface
+func (wb *Kathrein) Powers() (float64, float64, float64, error) {
+	return wb.getPhaseValues(kathreinRegPowers, 1)
+}
+
+var _ api.ChargeTimer = (*Kathrein)(nil)
+
+// ChargeDuration implements the api.ChargeTimer interface
+func (wb *Kathrein) ChargeDuration() (time.Duration, error) {
+	b, err := wb.conn.ReadHoldingRegisters(kathreinRegChargingDuration, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return time.Duration(binary.BigEndian.Uint16(b)) * time.Second, nil
+}
+
+var _ api.ChargeRater = (*Kathrein)(nil)
+
+// ChargedEnergy implements the api.ChargeRater interface
+func (wb *Kathrein) ChargedEnergy() (float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(kathreinRegChargingEnergy, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(binary.BigEndian.Uint32(b)) / 1e3, err
+}
+
+var _ api.MeterEnergy = (*Kathrein)(nil)
+
+// TotalEnergy implements the api.MeterEnergy interface
+func (wb *Kathrein) TotalEnergy() (float64, error) {
+	b, err := wb.conn.ReadHoldingRegisters(kathreinRegTotalEnergy, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(math.Float32frombits(binary.BigEndian.Uint32(b))), err
+}
+
+var _ api.PhaseSwitcher = (*Kathrein)(nil)
+
+// Phases1p3p implements the api.PhaseSwitcher interface
+func (wb *Kathrein) Phases1p3p(phases int) error {
+	var u uint16 = 0x0007 // Three phase charging
+
+	if phases == 1 {
+		u = 0x0001 // One phase charging
+	}
+
+	enabled, err := wb.Enabled()
+	if err == nil && enabled {
+		if err = wb.Enable(false); err != nil {
+			return err
+		}
+	}
+
+	// Switch phases
+	_, err = wb.conn.WriteSingleRegister(kathreinRegSetpointRelais, u)
+
+	// Re-enable charging if it was previously enabled
+	if err == nil && enabled {
+		err = wb.Enable(true)
+	}
+
+	return err
+}
+
+var _ api.PhaseGetter = (*Kathrein)(nil)
+
+// GetPhases implements the api.PhaseGetter interface
+func (wb *Kathrein) GetPhases() (int, error) {
+	b, err := wb.conn.ReadHoldingRegisters(kathreinRegSetpointRelais, 1)
+
+	if err != nil {
+		return 0, err
+	}
+
+	switch s := binary.BigEndian.Uint16(b); s {
+	case 0x0001,
+		0x0002,
+		0x0003: // Single phase
+		return 1, nil
+	case 0x0007:
+		return 3, nil
+	default:
+		return 0, err
```

Comment:
Gibts diesen Fall?

Und: err ist hier immer nil!

Reply:
It may happen? This may prevent simply catching any value, like Kathrein implementing a new return value (like 2phase switching?) in the future without breaking the evcc since any new value will be graciously ignored by evcc. Granted, err is "nil". This will be changed.